### PR TITLE
ENH: add timedelta as valid type for interpolate with method='time'

### LINKED ISSUE #6424: interpolate w/ method=time does not work with timedeltas
```
In [98]: pd.DataFrame({'v':[1,2,3]},index=pd.to_timedelta([1,2,3], unit="d")).interpolate(method="time")

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-98-9c399944f329> in <module>()
----> 1 pd.DataFrame({'v':[1,2,3]},index=pd.to_timedelta([1,2,3], unit="d")).interpolate(method="time")

/usr/lib64/python2.7/site-packages/pandas/core/generic.py in interpolate(self, method, axis, limit, inplace, downcast, **kwargs)
   2530                                           inplace=inplace,
   2531                                           downcast=downcast,
-> 2532                                           **kwargs)
   2533 
   2534         if inplace:

/usr/lib64/python2.7/site-packages/pandas/core/internals.py in interpolate(self, *args, **kwargs)
   2402 
   2403     def interpolate(self, *args, **kwargs):
-> 2404         return self.apply('interpolate', *args, **kwargs)
   2405 
   2406     def shift(self, *args, **kwargs):

/usr/lib64/python2.7/site-packages/pandas/core/internals.py in apply(self, f, *args, **kwargs)
   2373 
   2374             else:
-> 2375                 applied = getattr(blk, f)(*args, **kwargs)
   2376 
   2377             if isinstance(applied, list):

/usr/lib64/python2.7/site-packages/pandas/core/internals.py in interpolate(self, method, axis, index, values, inplace, limit, fill_value, coerce, downcast, **kwargs)
    819                                      inplace=inplace,
    820                                      downcast=downcast,
--> 821                                      **kwargs)
    822 
    823         raise ValueError("invalid method '{0}' to interpolate.".format(method))

/usr/lib64/python2.7/site-packages/pandas/core/internals.py in _interpolate(self, method, index, values, fill_value, axis, limit, inplace, downcast, **kwargs)
    880 
    881         # interp each column independently
--> 882         interp_values = np.apply_along_axis(func, axis, data)
    883 
    884         blocks = [make_block(interp_values, self.items, self.ref_items,

/usr/lib64/python2.7/site-packages/numpy/lib/shape_base.pyc in apply_along_axis(func1d, axis, arr, *args)
     77     outshape = asarray(arr.shape).take(indlist)
     78     i.put(indlist, ind)
---> 79     res = func1d(arr[tuple(i.tolist())],*args)
     80     #  if res is a number, then we have a smaller output array
     81     if isscalar(res):

/usr/lib64/python2.7/site-packages/pandas/core/internals.py in func(x)
    877             return com.interpolate_1d(index, x, method=method, limit=limit,
    878                                       fill_value=fill_value,
--> 879                                       bounds_error=False, **kwargs)
    880 
    881         # interp each column independently

/usr/lib64/python2.7/site-packages/pandas/core/common.py in interpolate_1d(xvalues, yvalues, method, limit, fill_value, bounds_error, **kwargs)
   1362         if not getattr(xvalues, 'is_all_dates', None):
   1363         # if not issubclass(xvalues.dtype.type, np.datetime64):
-> 1364             raise ValueError('time-weighted interpolation only works '
   1365                              'on Series or DataFrames with a '
   1366                              'DatetimeIndex')

ValueError: time-weighted interpolation only works on Series or DataFrames with a DatetimeIndex
```

### PR DESCRIPTION
**Pull request by aileronajay**
- [x] closes #6424 
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [ ] whatsnew entry

### CODE REVIEW

**Comment by aileronajay** [comment]
this PR is currently a WIP, tests have not been added yet. This iteration is for initial reviews

**Comment by aileronajay** [comment]
@TomAugspurger the change that is currently in this PR will support timedelta as part of the index and the case presented in the issue

pd.DataFrame({'v':[1,np.nan,5]},index=pd.to_timedelta([1,2,3], unit="d")).interpolate(method="time")

will work

>>> pd.DataFrame({'v':[1,np.nan,5]},index=pd.to_timedelta([1,2,3], unit="d")).interpolate(method="time")
          v
1 days  1.0
2 days  3.0
3 days  5.0

Though i think what we expect here is that interpolate should work if the passed data is of type datetime or timedelta, something like this (below) should work (like how cut was enhanced). Is this the expectation?

>>> pd.DataFrame({'v':[np.datetime64('2012-01-01'),np.nan,np.datetime64('2012-01-03')]},index=[1,2,3]).interpolate()
           v
1 2012-01-01
2        NaT
3 2012-01-03

**Comment by aileronajay** [comment]
@jreback @jorisvandenbossche @TomAugspurger need pointers about expected functionality. I have added support for time delta as index when interpolating. Should we also be able to interpolate between input time data?

**Comment by TomAugspurger** [comment]
@aileronajay IIRC the original issue https://github.com/pandas-dev/pandas/issues/6424 was about interpolating when the index was timedeltas. We can support interpolating time delta data like in your second example as well. That's fine either in this PR or a separate one.

**Comment by aileronajay** [comment]
@TomAugspurger i will then add support for timedelta as index in this PR and create a separate PR for interpolating time data

**Comment by aileronajay** [comment]
are there other changes required in this PR? (like more tests?) @sinhrks @jreback @TomAugspurger @jorisvandenbossche

File: pandas/core/missing.py [hunk 1]
```diff
@@ -187,7 +187,8 @@ def _interp_limit(invalid, fw_limit, bw_limit):
         if method in ('values', 'index'):
             inds = np.asarray(xvalues)
             # hack for DatetimeIndex, #1646
-            if issubclass(inds.dtype.type, np.datetime64):
+            if (issubclass(inds.dtype.type, np.datetime64) or
+                    issubclass(inds.dtype.type, np.timedelta64)):
```
Comment (jreback) [inline pandas/core/missing.py]: this should use the is_datetime64_dtype functions instead

Comment (jreback) [inline pandas/core/missing.py]: or better yet the needs_i8_conversion function

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/missing.py (hunk 1 above)
Comment (aileronajay) [inline pandas/core/missing.py]: @jreback I have made this change now

**Comment by aileronajay** [comment]
@jreback i have made the change requested in the review, are there other changes that need to be taken care of?

File: pandas/tests/series/test_missing.py [hunk 2]
```diff
@@ -891,6 +891,16 @@ def test_spline_error(self):
         with tm.assertRaises(ValueError):
             s.interpolate(method='spline', order=0)
 
+    def test_interp_timedelta64(self):
+        # GH 6424
+        tm._skip_if_no_scipy()
```
Comment (jorisvandenbossche) [inline pandas/tests/series/test_missing.py]: Is scipy needed for this interpolate method?

**Review by jorisvandenbossche — COMMENTED** [review]
Can you add a small notice in the v0.20.0.txt whatsnew file (in the 'other enhancements' section)

File: pandas/core/missing.py [hunk 3]
```diff
@@ -187,7 +188,7 @@ def _interp_limit(invalid, fw_limit, bw_limit):
         if method in ('values', 'index'):
             inds = np.asarray(xvalues)
             # hack for DatetimeIndex, #1646
-            if issubclass(inds.dtype.type, np.datetime64):
+            if (needs_i8_conversion(inds.dtype.type)):
```
Comment (TomAugspurger) [inline pandas/core/missing.py]: You can remove the redundant outer pair of parenthesis.

File: pandas/tests/series/test_missing.py (hunk 2 above)
Comment (TomAugspurger) [inline pandas/tests/series/test_missing.py]: It shouldn't be.

Comment (aileronajay) [inline pandas/tests/series/test_missing.py]: the interpolate method being used over here is 'time', i dont think it requires scipy, i just had this notion that we ignore the interpolation tests if scipy is absent, on a more thorough check i found that is not the case

File: pandas/tests/series/test_missing.py [hunk 4]
```diff
@@ -891,6 +891,16 @@ def test_spline_error(self):
         with tm.assertRaises(ValueError):
             s.interpolate(method='spline', order=0)
 
+    def test_interp_timedelta64(self):
```
Comment (aileronajay) [inline pandas/tests/series/test_missing.py]: sure, would it be better if i add it as one more in the existing test or as a new test?

Comment (aileronajay) [inline pandas/tests/series/test_missing.py]: @TomAugspurger i have added non uniform spacing as a test case now

File: pandas/core/missing.py (hunk 3 above)
Comment (aileronajay) [inline pandas/core/missing.py]: @TomAugspurger i have addressed this review comment

File: pandas/tests/series/test_missing.py (hunk 2 above)
Comment (aileronajay) [inline pandas/tests/series/test_missing.py]: I have removed the skip if no scipy statement

**Comment by aileronajay** [comment]
@jreback I have made the suggested changes

File: doc/source/whatsnew/v0.20.0.txt [hunk 5]
```diff
@@ -59,6 +59,7 @@ Other enhancements
 - ``Series`` provides a ``to_excel`` method to output Excel files (:issue:`8825`)
 - The ``usecols`` argument in ``pd.read_csv`` now accepts a callable function as a value  (:issue:`14154`)
 - ``pd.DataFrame.plot`` now prints a title above each subplot if ``suplots=True`` and ``title`` is a list of strings (:issue:`14753`)
+-  The ``pd.Series.interpolate`` now supports timedelta as index type(:issue:`6424`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: with ``method='time'`` (also space before ``(:issue:...``

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
minor doc change. ping on green.

File: doc/source/whatsnew/v0.20.0.txt [hunk 6]
```diff
@@ -59,6 +59,7 @@ Other enhancements
 - ``Series`` provides a ``to_excel`` method to output Excel files (:issue:`8825`)
 - The ``usecols`` argument in ``pd.read_csv`` now accepts a callable function as a value  (:issue:`14154`)
 - ``pd.DataFrame.plot`` now prints a title above each subplot if ``suplots=True`` and ``title`` is a list of strings (:issue:`14753`)
+-  The ``pd.Series.interpolate`` now supports timedelta as index type with method=`time` (:issue:`6424`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: small edit, but I can do on merge:

remove the leading ``The`` and ``method='time'`` needs to be in double backticks.

``timedelta as *an* index type`` (add an)

Comment (aileronajay) [inline doc/source/whatsnew/v0.20.0.txt]: @jreback should i make these changes or will be able to do that during merge?

**Comment by aileronajay** [comment]
@jreback there was a clean run in travis, appveyor failed after running for an hour, there appears to be a problem with appveyor

**Comment by jorisvandenbossche** [comment]
@aileronajay Thanks!

**Pull request merged**